kep/61-real-load-aware-scheduling/61-real-load-aware-scheduling.md

```diff
@@ -0,0 +1,608 @@
+
+Utilization of each node:
+
+
+```
+Ux → (1/4)*100 = 25
+Uy → (2/4)*100 = 50
+Uz → (3/4)*100 = 75
+```
+
+
+The score of each node :
+
+
+```
+X → 50 + 25 =  75
+Y → 50 + 50 = 100
+Z → 75 - 50 = 25
+```
+
+
+In the algorithm above, 50% is the target utilization rate we want to achieve on all nodes. We can be less aggressive by reducing it to 40% so that it has much lesser chances of going over 50% during spikes or unexpected loads. 
+
+In the 2nd step of the algorithm, one variant uses the current pod's total CPU limits instead of requests, to have a stronger upper bound of expected utilization.
+
+The 50% threshold value for utilisation will be made configurable via plugin argument.
+
+**Algorithm Analysis**
+
+<img src="images/bfbp-graph.png" alt="bfbp-graph" width="600" height="400"/>
+
+
+The above is a plot of the piecewise function outlined in the algorithm. The key observations here are manifold:
+
+
+
+1. As the utilization goes from 0 to 50%, we pack pods on those nodes by favoring them.
+2. The nodes are penalized linearly as the utilization goes beyond 50%, to spread the pod amongst those "hot" nodes.
+3. The positive slope begins from 50 and not from 0 because the range of score is from 0-100, and we want to maximize our score output for nodes we wish to favor so that the score is more substantial and other plugins do not affect it much.
+4. There is a break in the graph with a high drop due to the "penalty" we have.
+
+#### Spreading with BestFitBinPack
+
+
+To meet the 2nd constraint in the problem statement, we utilize the existing PodTopologySpread (PTS) plugin provided by K8s. 
+This is to avoid baking in another spreading algorithm within BFBP when there is an existing in-tree plugin and conflict our 4th goal.  
+It was proved with experiments that using a dynamic weight-based approach (changing weights of PTS and BFBP in every scheduling cycle) is not a good idea, neither it is supported by Kubernetes. 
+However, if we choose a static large weight for PTS, say 100, we can meet our goals. 
+The problem with this approach is that we will end up downplaying the scores of other important default in-tree plugins like Image Locality, Interpod Affinity, etc. with their default weights of 1. 
+Also, with more plugins we plan to add in the future, managing weights will not be simple. So for BFBP to work well with PTS without modifying weights, 
+the following extended algorithm is proposed that is called when scheduling replicas of a service, replica-set, replication controller, etc.
+
+
+**Algorithm**
+
+
+
+1. For each node, calculate S = (1000*PTS + PTS*BFBP)
+2. Scale each S to be between 0-100
+3. Return the scores
+
+**Algorithm Analysis**
+
+1. The algorithm favors nodes with high PTS scores with 1000*PTS part. The weight 1000 is chosen to give preference to PTS  and to avoid the congestion of scores upon scaling down, my magnifying the scores 
```

Comment:
You are right, the reason for combining is explained below in bold:

"However, if we choose a static large weight for PTS, say 100, we can meet our goals. 
**The problem with this approach is that we will end up downplaying the scores of other important default in-tree plugins like Image Locality, Interpod Affinity, etc. with their default weights of 1**. 
**Also, with more plugins we plan to add in the future, managing weights will not be simple**. So for BFBP to work well with PTS without modifying weights, 
the following extended algorithm is proposed that is called when scheduling replicas of a service, replica-set, replication controller, etc."

Reply:
I think a large weight works well for "standalone" plugins which are kind of hard requirements. In our case spreading is not a hard requirement. If PTS scores of two nodes are nearly equal (say 70, 80), we want to favor nodes with better image locality (say 100, 50). If we use 1000 as PTS weight, we will end up selecting node with 80 score always, when we want to select node with 70 score (due to better image locality). This is what "downplaying" scores means above.

With multiple plugins and multiple weights, the interplay of scores gets complex and hard to understand/debug - this is what "managing weights will not be simple" means.

Reply:
The above goes with over goal of "To not affect the behavior of default score plugins unless necessary."

Reply:
I'm not sure I like this.

It's inevitable that Score plugins may conflict with each other, or one plugin's impact gets diluted by others. Scoring is best-efforts anyway, expecting a 100% predictable scoring result is almost impossible. Even if you mixed a default weight of PTS into BFBP, it may not always work as expected. I'd still suggest BFBP be implemented on its own. You can guide the users to configure the weight of PTS when using BFBP.
On the other hand, mixing PTS score plugin would make the BFBP plugin hard to maintain. The PTS score plugin is not guaranteed to be exposed as a library, and can change its function API at any time, which makes the development of BFBP into risk.

Reply:
I agree about the conflict - it is unavoidable to have 0 conflict. What is important to understand here is that we try to **minimise** the conflict where we can. This also goes along the idea of a fair scheduler with different plugins given the weight of 1 and none of them is favored more over others.

>>>The PTS score plugin is not guaranteed to be exposed as a library, and can change its function API at any time, which makes the development of BFBP into risk."

About this, yes it could be a problem - I thought the framework API is almost stable now :)

Reply:
@ahg-g @alculquicondor any thoughts about mixing a plugin into another?

Reply:
> The problem with this approach is that we will end up downplaying the scores of other important default in-tree plugins like Image Locality, Interpod Affinity, etc. with their default weights of 1.

If we just keep the plugins separate, with a weight of 2 (default), we wouldn't affect other plugins.

>  You can guide the users to configure the weight of PTS when using BFBP.

+1

Reply:
@alculquicondor I doubt if weight of 2 will solve our goals in 1.19. If it does then I agree the justification above is not valid. Let me try this out.

Reply:
topology spreading in 1.19 is significantly different from 1.18 kubernetes/kubernetes#90475

Reply:
@alculquicondor good job for your work on PTS. I found no problems when I tested it for spreading with new default weight (as compared to 1.18). Since this works with weight of 2 for simple use cases, I believe for complex topologies or with more number of nodes, we wouldn't have to use a "large weight" like earlier to achieve our goals. I have removed the extended algorithm from KEP.

Reply:
@Huang-Wei @alculquicondor please review/approve.

Reply:
SG, you don't need to specify the default weight for now. Once you got a reasonable default number, return and update the KEP.

Reply:
SG. If you found that a larger weight is needed (for most cases), just update the wordings in this KEP.